File: src/main/java/eu/rekawek/coffeegb/gui/RecentRoms.java
package eu.rekawek.coffeegb.gui;

import java.util.LinkedList;
import java.util.List;
import java.util.Properties;
import java.util.stream.Collectors;

public class RecentRoms {

    private static final String KEY_PREFIX = "recent_rom.";
    private static final int MAX_ROMS = 10;

    private final Properties properties;

    private final LinkedList<String> roms;

    public RecentRoms(Properties properties) {
        this.properties = properties;
        roms = new LinkedList<>();
        for (int i = 0; i < MAX_ROMS; i++) {
            String key = KEY_PREFIX + i;
            if (properties.containsKey(key)) {
                roms.add(properties.getProperty(key));
            }
        }
    }

    public List<String> getRoms() {
        return roms;
    }

    public void addRom(String rom) {
        roms.remove(rom);
        roms.addFirst(rom);
        while (roms.size() > MAX_ROMS) {
            roms.removeLast();
        }
        cleanProperties();
        setProperties();
    }

    private void cleanProperties() {
        List<String> keys = properties.keySet().stream().map(o -> (String) o).filter(k -> k.startsWith(KEY_PREFIX)).collect(Collectors.toList());
        for (String k : keys) {
            properties.remove(k);
        }
    }

    private void setProperties() {
        for (int i = 0; i < roms.size(); i++) {
            properties.setProperty(KEY_PREFIX + i, roms.get(i));
        }
    }
}

File: src/main/java/eu/rekawek/coffeegb/swing/gui/properties/RecentRoms.kt
package eu.rekawek.coffeegb.swing.gui.properties

import java.util.*
import java.util.stream.Collectors

class RecentRoms(private val emulatorProperties: EmulatorProperties) {
    private val roms = LinkedList<String>()

    private val properties = emulatorProperties.properties

    init {
        for (i in 0 until MAX_ROMS) {
            val key = ROM_KEY_PREFIX + i
            if (properties.containsKey(key)) {
                roms.add(properties.getProperty(key))
            }
        }
    }

    fun getRoms(): List<String> {
        return roms
    }

    fun addRom(rom: String) {
        roms.remove(rom)
        roms.addFirst(rom)
        while (roms.size > MAX_ROMS) {
            roms.removeLast()
        }
        cleanProperties()
        setProperties()
        emulatorProperties.saveProperties()
    }

    private fun cleanProperties() {
        val keys = properties.keys.stream().map { o: Any -> o as String }.filter { k: String -> k.startsWith(ROM_KEY_PREFIX) }.collect(Collectors.toList())
        for (k in keys) {
            properties.remove(k)
        }
    }

    private fun setProperties() {
        for (i in roms.indices) {
            properties.setProperty(ROM_KEY_PREFIX + i, roms[i])
        }
    }

    private companion object {
        const val ROM_KEY_PREFIX = "rom.recent."
        const val MAX_ROMS = 10
    }
}